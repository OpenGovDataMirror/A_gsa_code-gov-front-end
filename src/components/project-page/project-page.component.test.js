import React from 'react'
import { shallow } from 'enzyme'

import repo from 'mocks/repo'
import { testRenderText, testRenderList, testRenderEmpty } from 'testUtils/render'
import CustomLink from 'components/custom-link'
import ProjectPage from 'components/project-page/project-page.component'

const props = {
  repo,
  match: {
    params: {
      repoID: 'repo-id'
    }
  },
  updateProject: jest.fn()
}

let wrapper
let instance
describe('components - ProjectPage', () => {
  beforeEach(() => {
    wrapper = shallow(<ProjectPage {...props} />)
    instance = wrapper.instance()
  })

  it('should update the project if it is initalized with the wrong repo', () => {
    wrapper = shallow(<ProjectPage {...props} repo={{ ...repo, repoID: 'other-id' }} />)
    expect(props.updateProject).toBeCalledWith(props.match.params.repoID)
  })

  describe('lastModifiedDateHTML', () => {
    it('should render the last modified date', () => {
      const expected = new Date(props.repo.date.lastModified).toLocaleDateString('en-US')
      testRenderText(instance.lastModifiedDateHTML, expected)
    })

    it('should render nothing if no last modfified date', () => {
      wrapper.setProps({ repo: { ...repo, date: undefined } })
      testRenderEmpty(instance.lastModifiedDateHTML)
    })
  })

  describe('repoTags', () => {
    it('should render the `tags` as links', () => {
      testRenderList(instance.repoTags, CustomLink, props.repo.tags.length)
    })
  })

  describe('usageType', () => {
    it('should render `Open Source` text if the usage type is open source', () => {
      wrapper.setProps({ repo: { ...repo, permissions: { usageType: 'openSource' } } })
      testRenderText(instance.usageType, 'Open Source')
    })

    it('should render `Government-Wide Reuse` text if the usage type is gov wide reuse', () => {
      wrapper.setProps({ repo: { ...repo, permissions: { usageType: 'governmentWideReuse' } } })
      testRenderText(instance.usageType, 'Government-Wide Reuse')
    })

    it('should render nothing if the usage type is neither open source nor gov wide reuse', () => {
      testRenderEmpty(instance.usageType)
    })
  })

  describe('license', () => {
    it('should render the license', () => {
      testRenderText(instance.license, repo.permissions.licenses[0].name)
    })

    it('should render nothing if license is not provided', () => {
      wrapper.setProps({ repo: { ...repo, permissions: undefined } })
      testRenderEmpty(instance.license)
    })
  })

  describe('displayLaborHours', () => {
    it('should render the labor hours', () => {
      testRenderText(instance.displayLaborHours, repo.contact.laborHours)
    })

    it('should render nothing if labor hours is not provided', () => {
      wrapper.setProps({ repo: { ...repo, contact: undefined } })
      testRenderEmpty(instance.displayLaborHours)
    })
  })

  describe('languages', () => {
    it('should render the `languages`', () => {
      const expected = repo.languages.join(', ')
      testRenderText(instance.languages, expected)
    })

    it('should render nothing if `languages` is not provided', () => {
      wrapper.setProps({ repo: { ...repo, languages: undefined } })
      testRenderEmpty(instance.languages)
    })
  })

  describe('contact', () => {
    it('should render the `email`', () => {
      testRenderText(instance.contact, props.repo.contact.email)
    })

    it('should render nothing if `email` is not provided', () => {
      wrapper.setProps({ repo: { ...repo, contact: undefined } })
      testRenderEmpty(instance.contact)
    })
  })

  describe('repositoryURL', () => {
    it('should render the repo url', () => {
      testRenderText(instance.repositoryURL, 'Visit Repo')
    })

    it('should render nothing if `email` is not provided', () => {
      wrapper.setProps({ repo: { ...repo, repositoryURL: undefined } })
      testRenderEmpty(instance.repositoryURL)
    })

    it('should render nothing if `repositoryURL` is a null string', () => {
      wrapper.setProps({ repo: { ...repo, repositoryURL: 'null' } })
      testRenderEmpty(instance.repositoryURL)
    })
  })

  describe('additionalData', () => {
    it('should render the additional repo data', () => {
      const expected = Object.keys(props.repo.additional_data).length
      testRenderList(instance.additionalData, '.metadata-value', expected)
    })

    it('should render nothing if additional data exists but has no values', () => {
      wrapper.setProps({ repo: { ...repo, additional_data: {} } })
      testRenderEmpty(instance.additionalData)
    })

    it('should render nothing if additional data is not provided', () => {
      wrapper.setProps({ repo: { ...repo, additional_data: undefined } })
      testRenderEmpty(instance.additionalData)
    })
  })

  describe('render', () => {
    it('should render correctly', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it('should render a loading indicator if no repo', () => {
      wrapper.setProps({ repo: undefined })
      expect(wrapper.text()).toMatch(/loading/i)
    })
  })
})
