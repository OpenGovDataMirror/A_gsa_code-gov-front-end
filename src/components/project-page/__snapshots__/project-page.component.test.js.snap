// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components - ProjectPage render should render correctly 1`] = `
<main
  id="main-content"
>
  <Connect(SiteBanner)
    title="Projects"
  />
  <Breadcrumbs
    crumbs={
      Array [
        Object {
          "text": "Home",
          "to": "/",
        },
        Object {
          "text": "repo-agency-acronym",
          "to": "/browse-projects?agencies=repo-agency-acronym",
        },
        Object {
          "text": "repo-name",
        },
      ]
    }
  />
  <section
    className="grid-container"
  >
    <header>
      <div
        className="margin-top-1"
      >
        <h2
          className="font-heading-xl"
        >
          repo-name
        </h2>
        <p
          className="font-body-2xs margin-y-2px"
        >
          Last updated on 12/10/2020
        </p>
        <div
          className="grid-row margin-top-0"
          id="repo-tags"
        >
          <Connect(CustomLinkComponent)
            className="usa-button font-body-2xs padding-1 padding-x-205 bg-primary text-lowercase radius-md margin-top-1 width-auto"
            key="repo-tag-1"
            to="/search?query=repo-tag-1"
          >
            <span
              className="text-white text-no-underline text-normal"
            >
              repo-tag-1
            </span>
          </Connect(CustomLinkComponent)>
          <Connect(CustomLinkComponent)
            className="usa-button font-body-2xs padding-1 padding-x-205 bg-primary text-lowercase radius-md margin-top-1 width-auto"
            key="repo-tag-2"
            to="/search?query=repo-tag-2"
          >
            <span
              className="text-white text-no-underline text-normal"
            >
              repo-tag-2
            </span>
          </Connect(CustomLinkComponent)>
        </div>
        <ul
          className="padding-0 font-body-2xs"
        >
          <li>
            <i
              className="icon icon-certificate padding-right-05"
            />
            repo-license
          </li>
          <li>
            <i
              className="icon icon-hourglass-end padding-right-05"
            />
            123 hours
          </li>
          <li>
            <i
              className="icon icon-code padding-right-05"
            />
            repo-lang-1, repo-lang-2
          </li>
          <li>
            <i
              className="icon icon-mail padding-right-05"
            />
            <a
              href="mailto:repo-email?Subject=Contribution%20Inquiry"
              target="_top"
            >
              repo-email
            </a>
          </li>
        </ul>
        <p
          className="maxw-none font-body-2xs"
        >
          repo-desc
        </p>
        <div>
          <br />
          <div
            className="metadata-entry"
          >
            <div
              className="metadata-key"
            >
              array
            </div>
            <div
              className="metadata-value"
            >
              <div
                key="array-repo-meta-arr-1"
              >
                repo-meta-arr-1
              </div>
            </div>
          </div>
          <div
            className="metadata-entry"
          >
            <div
              className="metadata-key"
            >
              url
            </div>
            <div
              className="metadata-value"
            >
              <a
                href="http://www.repo-meta-url.com"
                rel="noopener noreferrer"
                target="_blank"
              >
                <div>
                  http://www.repo-meta-url.com
                </div>
              </a>
            </div>
          </div>
          <div
            className="metadata-entry"
          >
            <div
              className="metadata-key"
            >
              string
            </div>
            <div
              className="metadata-value"
            >
              <div>
                repo-meta-string
              </div>
            </div>
          </div>
        </div>
        <br />
        <a
          className="usa-button margin-bottom-6 margin-top-neg-2px"
          href="https://github.com/repo-url"
          rel="noopener noreferrer"
          target="_blank"
        >
          Visit Repo
        </a>
      </div>
    </header>
  </section>
</main>
`;
